File: src/main/java/ds25/hotel/reservation/management/system/domain/user/application/AuthService.java
package ds25.hotel.reservation.management.system.domain.user.application;

import ds25.hotel.reservation.management.system.config.AppConfig;
import ds25.hotel.reservation.management.system.di.Service;
import ds25.hotel.reservation.management.system.domain.user.User;
import ds25.hotel.reservation.management.system.domain.user.UserRepository;
import ds25.hotel.reservation.management.system.ui.auth.data.LoginForm;
import ds25.hotel.reservation.management.system.ui.auth.data.RegisterForm;
import ds25.hotel.reservation.management.system.ui.dto.user.UserDto;
import ds25.hotel.reservation.management.system.util.PasswordEncryptor;
import jakarta.inject.Inject;
import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validator;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@Service
public class AuthService {
	final private UserRepository userRepository;
	final private PasswordEncryptor passwordEncryptor;
	final private Validator validator;

	@Inject
	public AuthService(UserRepository userRepository, PasswordEncryptor passwordEncryptor, Validator validator) {
		this.userRepository = userRepository;
		this.passwordEncryptor = passwordEncryptor;
		this.validator = validator;
	}

	/**
	 * 로그인 메소드
	 *
	 * @param loginForm 로그인 폼
	 * @return 로그인 성공시 유저 정보, 실패시 null
	 * @author 김남주
	 */
	public UserDto login(LoginForm loginForm) throws Exception {
		log.info("User Log in method is called");

		Optional<User> user = userRepository.findByEmail(loginForm.email());
		if (user.isEmpty() || !passwordEncryptor.matches(loginForm.password(), user.get().getPassword())) {
			throw new Exception("Invalid email or password");
		}

		AppConfig.getInstance().setUser(UserDto.from(user.get()));

		return UserDto.from(user.get());
	}

	public boolean isExistEmail(String email) {
		log.info("isExistId method called");
		return userRepository.existsByEmail(email);
	}

	/**
	 * 회원가입 메소드
	 *
	 * @param registerForm 유저 정보
	 * @return 회원가입 성공시 유저 정보, 실패시 null
	 * @author 김남주
	 */
	public UserDto register(RegisterForm registerForm) throws RuntimeException {
		log.info("User Register method is called");

		Set<ConstraintViolation<RegisterForm>> violations = validator.validate(registerForm);
		if (!violations.isEmpty()) {
			String messages = violations.stream().map(ConstraintViolation::getMessage).collect(Collectors.joining());
			throw new RuntimeException(messages);
		}

		User newUser = userRepository.save(User.builder()
				.email(registerForm.email())
				.name(registerForm.name())
				.password(registerForm.password())
				.phone(registerForm.phone())
				.build());

		return UserDto.from(newUser);
	}

	/**
	 * 로그아웃 하는 메소드
	 *
	 * @author 김남주
	 */
	public void logout() {
		log.info("User logout method called");
		AppConfig.getInstance().setUser(null);
	}

}

File: src/main/java/ds25/hotel/reservation/management/system/domain/user/application/AuthService.kt
package ds25.hotel.reservation.management.system.domain.user.application

import ds25.hotel.reservation.management.system.global.config.AuthContext
import ds25.hotel.reservation.management.system.domain.user.dao.UserRepository
import ds25.hotel.reservation.management.system.domain.user.domain.User
import ds25.hotel.reservation.management.system.domain.user.dto.LoginRequest
import ds25.hotel.reservation.management.system.domain.user.dto.SignUpRequest
import ds25.hotel.reservation.management.system.domain.user.exception.AuthException
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class AuthService(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder
) {
    /**
     * 로그인
     * @param loginRequest 로그인 요청
     * @return 로그인한 사용자
     */
    @Transactional(readOnly = true)
    fun login(loginRequest: LoginRequest): User {
        // 사용자 조회
        val user: User = userRepository.findByEmail(loginRequest.email)
            ?: throw AuthException("User not found")

        // 비밀번호 확인
        if (!passwordEncoder.matches(loginRequest.password, user.password)) {
            throw AuthException("Password is incorrect")
        }

        AuthContext.setUser(user)

        return user
    }

    /**
     * 로그아웃
     */
    fun logout() {
        AuthContext.clear()
    }

    /**
     * 회원가입
     * @param user 사용자
     * @return 가입한 사용자
     */
    @Transactional
    fun signUp(signUpRequest: SignUpRequest): User {
        // 이메일 중복 확인
        if (userRepository.existsByEmail(signUpRequest.email)) {
            throw AuthException("Email is already in use")
        }
        // 사용자 생성
        val user = User(signUpRequest, passwordEncoder)
        println("user: $user")

        return userRepository.save(user)
    }

    /**
     * 아이디 사용가능 여부 확인
     * @param email 이메일
     * @return 사용가능 여부 (true: 사용가능, false: 사용불가)
     */
    @Transactional(readOnly = true)
    fun isAvailableEmail(email: String): Boolean {
        return !userRepository.existsByEmail(email)
    }
}
